### PR TITLE
Include book status when getting books

### PR DESCRIPTION
**Pull request by natasatutuspd**

### CODE REVIEW

File: src/main/java/com/productdock/library/search/book/BookMapper.java [hunk 1]
```diff
@@ -1,16 +1,75 @@
...
+        return bookStatusWrapper.getRecords();
+    }
+
+    private int getAvailableBooksCount(BookDocument bookDocument) {
+        var bookStatusWrapper = bookDocument.getBookStatusWrapper();
+        if (bookStatusWrapper == null) {
+            return 0;
+        }
+        return bookStatusWrapper.getAvailableBooksCount();
+    }
```
Comment (NenadBecanovicPD) [inline src/main/java/com/productdock/library/search/book/BookMapper.java:53]: Maybe move methods to BookDocument class?

File: src/main/java/com/productdock/library/search/book/BookMapper.java [hunk 2]
```diff
@@ -1,16 +1,75 @@
...
-public interface BookMapper {
+@Component
+public class BookMapper {
 
-    BookDto toBookDto(BookDocument bookDocument);
+    @Autowired
+    private RecordDtoMapper recordDtoMapper;
 
-    BookDocument toBookDocument(InsertBookMessage insertBookMessage);
+    public BookDto toBookDto(BookDocument bookDocument) {
```
Comment (milicazivkov85) [inline src/main/java/com/productdock/library/search/book/BookMapper.java:21]: Extract till you drop:

In the end, `toBookDto()` should be simple as:
```
        var bookDto = new BookDto();
        mapSimpleProperties(bookDto, bookDocument);
        mapRecords(bookDto, bookDocument.getRecords());
        return bookDto;
```

Try to find other opportunities for extraction.

File: src/test/java/com/productdock/library/search/book/BookMapperShould.java [hunk 3]
```diff
@@ -21,27 +25,86 @@ class BookMapperShould {
     @Test
     void mapBookDocumentToBookDto() {
         var bookDocument = defaultBookDocumentBuilder()
-                .id("123").title("Book title").author("Book author").cover("Book cover").build();
+                .bookId("123").title("Book title").author("Book author").cover("Book cover").build();
+        List<Record> records = new ArrayList<>();
```
Comment (milicazivkov85) [inline src/test/java/com/productdock/library/search/book/BookMapperShould.java:29]: Setting up test object (BookDocument in this case) should be cleaner. Try using Builder pattern to set up entire BookDocument.

**Review by milicazivkov85 — CHANGES_REQUESTED** [review]

File: src/main/java/com/productdock/library/search/book/BookMapper.java [hunk 4]
```diff
@@ -2,51 +2,46 @@
...
-        var records = getBookDocumentRecords(bookDocument);
-        int availableBookCount = getAvailableBooksCount(bookDocument);
-        for (int i = 1; i <= availableBookCount; i++) {
-            var bookRecord = BookDocument.RentalState.Record.builder().status(BookStatus.AVAILABLE).build();
-            records.add(bookRecord);
-        }
+    private void mapRecords(BookDocument bookDocument, BookDto bookDto) {
+        var records = bookDocument.getRentalState().getRecords();
+        var availableBookCount = bookDocument.getRentalState().getAvailableBooksCount();
+        records.addAll(createAvailableRecords(availableBookCount));
```
Comment (milicazivkov85) [inline src/main/java/com/productdock/library/search/book/BookMapper.java:28]: toBookDto method violates Have no side effect rule. It alters the records collection belonging to the input parameter (BookDocument). 
You can fix this by changed the way list is combined:
```
        var availableRecords = createAvailableRecords(availableBookCount);
        var allRecords = Collections.join(records, availableRecords);
```
That way bookDocument.getRentalState().getRecords() will stay unchanged , and BookDto will still get combined records.

**Review by milicazivkov85 — APPROVED** [review]
It is approved, but please address this minor issues before merging.
I have added one additional commit, to make cleaner mother objects, that will allow use of fluent builder. See last commit. Test are failing because of the issue with toBookDto (it has side effects) but as soon as you fix it it will go green again.

**Pull request merged**